File: TOP150Review/Longest Substring Without Repeating Characters.py

# sliding windows solution 
class Solution:
    
    def lengthOfLongestSubstring(self, s: str) -> int:
        # 用window去keep track在範圍內不重複的字 
        window = set()    
        sol = 0 
        left = 0 
        right = 0 
        size = len(s) 
        
        # 開放區間的 sliding window [left , right)
        while right < size : 
            
            # 將新的字加入區間 , 這邊在insert的時候就要檢查有沒有元素重複
            add_char = s[right]   

            #如果有元素已經在window內 , 移動left指標 , 直到add_char不在window內部 
            if add_char in window :
                
                while left < right and  add_char in window :
                    window.remove( s[left] ) 
                    left += 1

            window.add(add_char)  
            right += 1 
            
            sol = max(sol , right-left)
        
        
        return sol 
            

            

File: RUSH/Longest Substring Without Repeating Characters.js
/**
 * @param {string} s
 * @return {number}
 */


// sliding window
var lengthOfLongestSubstring = function(s) {
    
    let window = {} ; 
    let left = 0 , right = 0  ; 
    let solution = 0 ; 


    while (right < s.length){

        // add new element into window 
        let addElement = s[right] ; 
        let repeat = false ; 
        if (window[addElement]){
            window[addElement] += 1 
            repeat = true 
        }
        else{
            window[addElement] = 1 
        }
        right += 1 ; 

        while (repeat){
            
            let removeElement = s[left] ; 
            window[removeElement] -= 1 
            if (window[removeElement] === 0){
                delete window[removeElement] 
            }
            else if(window[removeElement] === 1){
                repeat = false ; 
            }
            left += 1 ; 

        }

        solution = Math.max(solution , right-left)

    }
    return solution ; 

};

// sliding window
var lengthOfLongestSubstring = function(s) {
    let window = new Set() ; 
    let left = 0 , right = 0 ; 
    let solution = 0 ; 

    while (right < s.length){

        let addElement = s[right] ; 
        let repeatElement = null ; 
        if (window.has(addElement)){
            repeatElement = addElement ; 
        }
        else{
            window.add(addElement)
        }

        while (repeatElement){
            let removeElement = s[left] ; 
            if (removeElement === repeatElement){
                repeatElement = null ; 
            }
            else {
                window.delete(removeElement) ; 
            }
            left += 1 ;
        }
        right+=1 ; 
        solution = Math.max(solution , right - left ) ; 
    }
    return solution ; 

}